File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/language/Transformable.java
package com.raizlabs.android.dbflow.sql.language;

import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.sql.language.property.IProperty;

import java.util.List;

/**
 * Description: Provides a standard set of methods for ending a SQLite query method. These include
 * groupby, orderby, having, limit and offset.
 */
public interface Transformable<T> {

    @NonNull
    Where<T> groupBy(NameAlias... nameAliases);

    @NonNull
    Where<T> groupBy(IProperty... properties);

    @NonNull
    Where<T> orderBy(@NonNull NameAlias nameAlias, boolean ascending);

    @NonNull
    Where<T> orderBy(@NonNull IProperty property, boolean ascending);

    @NonNull
    Where<T> orderBy(@NonNull OrderBy orderBy);

    @NonNull
    Where<T> limit(int count);

    @NonNull
    Where<T> offset(int offset);

    @NonNull
    Where<T> having(SQLOperator... conditions);

    @NonNull
    Where<T> orderByAll(@NonNull List<OrderBy> orderBies);
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/language/Transformable.kt
package com.raizlabs.android.dbflow.sql.language

import com.raizlabs.android.dbflow.sql.language.property.IProperty

/**
 * Description: Provides a standard set of methods for ending a SQLite query method. These include
 * groupby, orderby, having, limit and offset.
 */
interface Transformable<T : Any> {

    fun groupBy(vararg nameAliases: NameAlias): Where<T>

    fun groupBy(vararg properties: IProperty<*>): Where<T>

    fun orderBy(nameAlias: NameAlias, ascending: Boolean): Where<T>

    fun orderBy(property: IProperty<*>, ascending: Boolean): Where<T>

    fun orderBy(orderBy: OrderBy): Where<T>

    fun limit(count: Int): Where<T>

    fun offset(offset: Int): Where<T>

    fun having(vararg conditions: SQLOperator): Where<T>

    fun orderByAll(orderBies: List<OrderBy>): Where<T>
}


infix fun <T : Any> Transformable<T>.groupBy(nameAlias: NameAlias): Where<T> = groupBy(nameAlias)

infix fun <T : Any> Transformable<T>.groupBy(property: IProperty<*>): Where<T> = groupBy(property)

infix fun <T : Any> Transformable<T>.orderBy(orderBy: OrderBy): Where<T> = orderBy(orderBy)

infix fun <T : Any> Transformable<T>.limit(limit: Int): Where<T> = limit(limit)

infix fun <T : Any> Transformable<T>.offset(offset: Int): Where<T> = offset(offset)

infix fun <T : Any> Transformable<T>.having(sqlOperator: SQLOperator): Where<T> = having(sqlOperator)